uprotocol/communication/rpcresult.py

```diff
@@ -0,0 +1,171 @@
+        pass
+
+    @abstractmethod
+    def filter(self, f: Callable[[T], bool]) -> "RpcResult":
+        pass
+
+    @abstractmethod
+    def failure_value(self) -> UStatus:
+        pass
+
+    @abstractmethod
+    def success_value(self) -> T:
+        pass
+
+    @staticmethod
+    def success(value: T) -> "RpcResult":
+        return Success(value)
+
+    @staticmethod
+    def failure(
+        value: Union[
+            UStatus,
+            "Failure",
+            Exception,
+        ] = None,
+        code: UCode = UCode.UNKNOWN,
+        message: str = "",
+    ) -> "RpcResult":
+        return Failure(value, code, message)
+
+    @staticmethod
+    def flatten(result: "RpcResult") -> "RpcResult":
+        return result.flat_map(lambda x: x)
+
+
+class Success(RpcResult):
+    def __init__(self, value: T):
+        self.value = value
+
+    def is_success(self) -> bool:
+        return True
+
+    def is_failure(self) -> bool:
+        return False
+
+    def get_or_else(self, default_value: Callable[[], T]) -> T:
+        return self.success_value()
+
+    def map(self, f: Callable[[T], T]) -> RpcResult:
+        try:
+            return self.success(f(self.success_value()))
+        except Exception as e:
+            return self.failure(e)
+
+    def flat_map(self, f: Callable[[T], RpcResult]) -> RpcResult:
+        try:
+            return f(self.success_value())
+        except Exception as e:
+            return self.failure(e)
+
+    def filter(self, f: Callable[[T], bool]) -> RpcResult:
+        try:
+            return (
+                self
+                if f(self.success_value())
+                else self.failure(code=UCode.FAILED_PRECONDITION, message="filtered out")
+            )
+        except Exception as e:
+            return self.failure(e)
+
+    def failure_value(self) -> UStatus:
+        raise ValueError("Method failure_value() called on a Success instance")
+
+    def success_value(self) -> T:
+        return self.value
+
+    def __str__(self) -> str:
+        return f"Success({self.success_value()})"
+
+
+class Failure(RpcResult):
+    def __init__(
+        self,
+        value: Union[UStatus, "Failure", Exception, None] = None,
+        code: UCode = UCode.UNKNOWN,
+        message: str = "",
+    ):
+        if isinstance(value, UStatus):
+            self.value = value
+        elif isinstance(value, Exception):
+            self.value = UStatus(code=code, message=str(value))
+        elif isinstance(value, Failure):
+            self.value = value.failure_value()
+        else:
+            self.value = UStatus(code=code, message=message)
+
+    def is_success(self) -> bool:
+        return False
+
+    def is_failure(self) -> bool:
+        return True
+
+    def get_or_else(self, default_value: Callable[[], T]) -> T:
+        if callable(default_value):
+            return default_value()
+        return default_value
+
+    def map(self, f: Callable[[T], T]) -> RpcResult:
```

Comment:
are these needed? I know they are in Java so we can use them in a stream but are they needed in Python?

Reply:
I will remove the unused parts in the next PR after checking the simulator code.